src/library_async.js

```diff
@@ -238,6 +238,7 @@ mergeInto(LibraryManager.library, {
     saveStack: '',
     yieldCallbacks: [],
     postAsync: null,
+    restartFunc: null,
```

Comment:
The compiler generates code like this for every emterpreted function:

```js
function _stringf($0) {
...
 emterpret(eb + 29584 | 0);
 return HEAP32[EMTSTACKTOP >> 2] | 0;
}
function _floatf() {
...
 emterpret(eb + 30444 | 0);
 return +HEAPF64[EMTSTACKTOP >> 3];
}
```
If we want to get a return value after calling `emterpret`, we need to know the return type. The easiest way I found is to just store this function on `EmterpreterAsync` so that the async handler can call it and get the return value. Or do you have a better idea?
Should I add a comment in the code?

Reply:
It seems like you add both the `restartFunc` and a finalizer, and pass the value returned from one to the other? It seems like all we need is the finalizer, perhaps, and move the other logic to there?

But I may be missing something, I still don't feel like I understand the stack operations here.

Reply:
`restartFunc` may be called multiple times during an async ccall, but the finalizers are called just once at the end. So we can't move that logic into the finalizer.
The only purpose of `restartFunc` is to get the return value. `emterpret` places the return value onto the stack. If we want to read the value from the stack, we need to know its type. But the only place in the JS code generated by the compiler is, as far as I know, in these functions like the ones I commented (the documentation calls them 'trampolines'). Inside the trampoline, the emterpreter is called and then the return value is taken from the stack with the appropriate type, and returned.

If you really don't like the `restartFunc` solution, we could also generate some additional data structure from which we can tell the return type. We could then look it up inside ccall and do something like:
```js
if (return type is int32) {
  ret = HEAP32[EMTSTACKTOP >> 2] | 0;
} else {
  ret = +HEAPF64[EMTSTACKTOP >> 3];
}
```
But that means generating additional code for every emterpreted function, whereas my solution only adds a constant amount of code.